### PR TITLE
Add support for Opensearch_Dashboards

### PR DESCRIPTION
**Pull request by raihankhan**
Signed-off-by: raihankhan <raihan@appscode.com>

### CODE REVIEW

File: pkg/distribution/open_search/config.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+		Name:      ed.dashboard.ServiceName(),
+		Namespace: ed.dashboard.Namespace,
+	}, &edSvc)
+
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to get dashboard client service: %s", ed.dashboard.ServiceName()))
+	}
+
+	secretDataKey := dapi.OpensearchDasboardsConfigFileName
+	dashboardServerPort := edSvc.Spec.Ports[0].Port
```
Comment (kamolhasan) [inline pkg/distribution/open_search/config.go:48]: Check port before accessing the index

File: pkg/distribution/open_search/config.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+			Name:      authSecretName,
+		}, &authSecret); err != nil {
+			return errors.Wrap(err, fmt.Sprintf("failed to get authsecret: %s/%s", ed.dashboard.Namespace, authSecretName))
+		}
+
+		secretData[dapi.OS_USER_KEY] = string(authSecret.Data[core.BasicAuthUsernameKey])
+		secretData[dapi.OS_PASSWORD_KEY] = string(authSecret.Data[core.BasicAuthPasswordKey])
+
+		if ed.db.Spec.EnableSSL {
+			//databaseCA := ed.dashboard.CertificateFilePath(dapi.ElasticsearchDashboardKibanaConfigDir, databaseCertAlias, dapi.CaCertKey)
```
Comment (kamolhasan) [inline pkg/distribution/open_search/config.go:105]: remove unwanted comments

**Review by kamolhasan — CHANGES_REQUESTED** [review]

File: pkg/distribution/open_search/config.go (hunk 1 above)
Comment (raihankhan) [reply pkg/distribution/open_search/config.go]: added checking

File: pkg/distribution/open_search/config.go (hunk 2 above)
Comment (raihankhan) [reply pkg/distribution/open_search/config.go:116]: removed

File: pkg/controllers/dashboard/reconciler.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+	// Ensure deployments are created
+	// Cert secrets volumes are mounted
+	// Credentials are provided as ENV variable
+	err = ed.ReconcileDeployment(ctx)
+	if err != nil {
+		return ed.UpdatedElasticsearchDashboard(), errors.Wrap(err, "failed to reconcile deployment")
+	}
+
+	return ed.UpdatedElasticsearchDashboard(), err
```
Comment (kamolhasan) [inline pkg/controllers/dashboard/reconciler.go:48]: Remove if not used

File: pkg/distribution/open_search/config.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+	}
+
+	// marshal the data to json->yaml
+	// pass the data as default config data to secret
+	data, err := yaml.Marshal(secretData)
+	if err != nil {
+		return errors.Wrap(err, "failed to create config secret data")
+	}
+
+	maxOldSpace := (ed.dashboard.Spec.PodTemplate.Spec.Resources.Requests.Memory().Value() * 3) / (4 * 1024 * 1024)
```
Comment (kamolhasan) [inline pkg/distribution/open_search/config.go:148]: Add comments, references

**Review by kamolhasan — CHANGES_REQUESTED** [review]

File: pkg/controllers/dashboard/controller.go [hunk 5]
```diff
@@ -79,159 +79,80 @@ func (r *ElasticsearchDashboardReconciler) Reconcile(ctx context.Context, req ct
...
 
 	// Ensure finalizers are added to dashboard
-	// removed if dashboard is on deletion
-	err = r.ensureFinalizers(dashboard, ctx)
-
+	// remove if dashboard is being deleted
+	err = r.ensureFinalizers(ctx, dashboard)
 	if err != nil {
+		if err.Error() == dapi.DashboardOnDeletion {
+			r.Log.Error(err, fmt.Sprintf("Dashboard %s/%s is on deletion", dashboard.Namespace, dashboard.Name))
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/controllers/dashboard/controller.go:86]: changed to info

File: pkg/controllers/dashboard/controller.go [hunk 6]
```diff
@@ -79,159 +79,80 @@ func (r *ElasticsearchDashboardReconciler) Reconcile(ctx context.Context, req ct
...
-	// no need to reconcile further
-	if !dashboard.ObjectMeta.DeletionTimestamp.IsZero() {
-		r.Log.Info(dashboard.Name + " is being deleted")
-		return ctrl.Result{}, err
-	}
+	// Update dashboard phase from current conditions
+	r.updatePhaseFromCondition(ctx, dashboard)
 
-	// Get database from databaseRef
+	// Get database using databaseRef
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/controllers/dashboard/controller.go:96]: reconstructed

File: pkg/controllers/dashboard/controller.go [hunk 7]
```diff
@@ -79,159 +79,80 @@ func (r *ElasticsearchDashboardReconciler) Reconcile(ctx context.Context, req ct
...
-	err = r.ensureCertSecrets(dashboard, ctx)
-	if err != nil {
-		r.Log.Error(err, "ensureCertSecrets failed")
-		return ctrl.Result{}, client.IgnoreNotFound(err)
-	}
-
 	// Ensure authSecrets exists
 	// when database is security enabled
-	err = r.ensureAuthSecret(dashboard, ctx)
+	err = r.ensureAuthSecret(ctx, dashboard)
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/controllers/dashboard/controller.go:122]: verification added

File: pkg/controllers/dashboard/controller.go [hunk 8]
```diff
@@ -79,159 +79,80 @@ func (r *ElasticsearchDashboardReconciler) Reconcile(ctx context.Context, req ct
...
 	if err != nil {
 		r.Log.Error(err, "EnsureAuthSecrets failed")
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	// Ensure database client certs exists
-	// This will be used to ensure secure data transmission between database and server
-	err = r.ensureDatabaseClientCert(dashboard, ctx)
+	// Ensure clusterIP service (Acts as client service for dashboard instance)
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/controllers/dashboard/controller.go:128]: fixed

File: pkg/controllers/dashboard/controller.go [hunk 9]
```diff
@@ -79,159 +79,80 @@ func (r *ElasticsearchDashboardReconciler) Reconcile(ctx context.Context, req ct
...
+	err = r.reconcileService(ctx, dashboard)
+
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/controllers/dashboard/controller.go:130]: removed

File: pkg/controllers/dashboard/controller.go [hunk 10]
```diff
@@ -269,3 +188,82 @@ func (r *ElasticsearchDashboardReconciler) updatePhaseFromCondition(dashboard *d
...
+	var err error
+	var deployment apps.Deployment
+
+	// Get deployment
+	if err = r.Client.Get(ctx, types.NamespacedName{
+		Namespace: dashboard.Namespace,
+		Name:      dashboard.DeploymentName(),
+	}, &deployment); err != nil {
+		if kerr.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/controllers/dashboard/controller.go:203]: fixed

File: pkg/controllers/dashboard/controller.go [hunk 11]
```diff
@@ -269,3 +188,82 @@ func (r *ElasticsearchDashboardReconciler) updatePhaseFromCondition(dashboard *d
...
+			in.Status.Conditions = kmapi.SetCondition(dashboard.Status.Conditions, kmapi.Condition{
+				Type:               string(dapi.DashboardConditionDeploymentReconciled),
+				Status:             core.ConditionTrue,
+				ObservedGeneration: dashboard.Generation,
+				Reason:             dapi.DashboardDeploymentAvailable,
+				Message:            fmt.Sprintf("Deployment: %s/%s is Available\n", dashboard.Namespace, dashboard.DeploymentName()),
+			})
+
+			// if deployment is available check service is also existing or not
+			if kmapi.HasCondition(dashboard.Status.Conditions, string(dapi.DashboardConditionServiceReconciled)) &&
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/controllers/dashboard/controller.go:225]: reconstructed checking algo

File: pkg/controllers/dashboard/health.go [hunk 12]
```diff
@@ -40,6 +40,11 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+type CertSecrets struct {
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/controllers/dashboard/health.go:43]: reconstructed

File: pkg/controllers/dashboard/reconciler.go (hunk 3 above)
Comment (raihankhan) [reply pkg/controllers/dashboard/reconciler.go:48]: removed

File: pkg/distribution/open_search/config.go (hunk 4 above)
Comment (raihankhan) [reply pkg/distribution/open_search/config.go:148]: added

File: pkg/controllers/dashboard/health.go [hunk 13]
```diff
@@ -180,31 +207,32 @@ func (r *ElasticsearchDashboardReconciler) CheckElasticsearchDashboardHealthOnce
...
+		// use server cert ca for rootca as issuer ref is not taken into account
+		clientCA := x509.NewCertPool()
+		rootCA := x509.NewCertPool()
 
-		dashboardCA, err := r.findSecret(dashboardCaName, dashboard, ctx)
-		if err != nil {
-			r.Log.Error(err, "failed to get databaseCaCert secret")
-			return nil, err
-		}
+		crt, _ := tls.X509KeyPair(certSecret.Data[core.TLSCertKey], certSecret.Data[core.TLSPrivateKeyKey])
```
Comment (kamolhasan) [inline pkg/controllers/dashboard/health.go:228]: Handle error

File: pkg/controllers/dashboard/health.go [hunk 14]
```diff
@@ -227,12 +255,12 @@ func (r *ElasticsearchDashboardReconciler) GetElasticsearchDashboardClient(dashb
...
 }
 
-func (r *ElasticsearchDashboardReconciler) GetDashboardClient(ctx context.Context, dashboard *dapi.ElasticsearchDashboard, certSecret ...CertSecrets) (*goed.EDClient, error) {
+func (r *ElasticsearchDashboardReconciler) GetDashboardClient(ctx context.Context, dashboard *dapi.ElasticsearchDashboard, clientOps *goed.DashboardClientOpts) (*goed.EDClient, error) {
 	clientConfig := goed.Config{
 		Host: getHostPath(dashboard),
 		Api:  dapi.KibanaStatusEndpoint,
 		Transport: &http.Transport{
-			IdleConnTimeout: time.Second * 3,
+			IdleConnTimeout: time.Second * 5,
```
Comment (kamolhasan) [inline pkg/controllers/dashboard/health.go:263]: Keep it 3 second

**Review by kamolhasan — CHANGES_REQUESTED** [review]

File: pkg/controllers/dashboard/controller.go [hunk 15]
```diff
@@ -79,159 +77,81 @@ func (r *ElasticsearchDashboardReconciler) Reconcile(ctx context.Context, req ct
...
 
-	// if dashboard is on deletion return from here
-	// no need to reconcile further
-	if !dashboard.ObjectMeta.DeletionTimestamp.IsZero() {
-		r.Log.Info(dashboard.Name + " is being deleted")
-		return ctrl.Result{}, err
+	// Get database using databaseRef
+	// it will be used to get version, disableSecurity, authSecret
+	r.db = &api.Elasticsearch{}
+	esName := dashboard.Spec.DatabaseRef.Name
```
Comment (kamolhasan) [inline pkg/controllers/dashboard/controller.go:97]: Check pointer

**Review by kamolhasan — APPROVED** [review]
LGTM!

**Pull request merged**